dsl_parser/import_resolver/abstract_import_resolver.py

```diff
@@ -59,29 +59,51 @@ def read_import(import_url):
-                        13, '{0} {1}; {2}'.format(
-                            error_str, import_url, err))
-                    raise ex
-                time.sleep(DEFAULT_RETRY_DELAY)
-                num_retries += 1
-            except requests.URLRequired as err:
-                ex = exceptions.DSLParsingLogicException(
-                    13, '{0} {1}; {2}'.format(
-                        error_str, import_url, err))
-                raise ex
+        number_of_attempts = MAX_NUMBER_RETRIES + 1
+
+        # Defines on which errors we should retry the import.
+        def _is_recoverable_error(e):
+            return isinstance(e, (requests.ConnectionError, requests.Timeout))
+
+        # Defines on which return values we should retry the import.
+        def _is_internal_error(result):
+            return hasattr(result, 'status_code') and result.status_code >= 500
+
+        @retry(stop_max_attempt_number=number_of_attempts,
+               wait_fixed=DEFAULT_RETRY_DELAY,
+               retry_on_exception=_is_recoverable_error,
+               retry_on_result=_is_internal_error)
+        def get_import():
+            response = requests.get(import_url,
+                                    timeout=DEFAULT_REQUEST_TIMEOUT)
+            # The response is a valid one, and the content should be returned
+            if 200 <= response.status_code < 300:
+                return response.text
+            # If the response status code is above 500, an internal server
+            # error has occurred. The return value would be caught by
+            # _is_internal_error (as specified in the decorator), and retried.
+            elif response.status_code >= 500:
+                return response
+            # Any other response should raise an exception.
             else:
-                if 200 <= response.status_code < 300:
-                    return response.text
-                else:
-                    ex = exceptions.DSLParsingLogicException(
-                        13, '{0} {1}; status code: {2}'.format(
-                            error_str, import_url, response.status_code))
-                    raise ex
+                invalid_url_err = exceptions.DSLParsingLogicException(
+                    13, '{0} {1}; status code: {2}'.format(
+                        error_str, import_url, response.status_code))
+                raise invalid_url_err
+
+        try:
+            import_result = get_import()
+            # If the error is an internal error only. A custom exception should
+            # be raised.
+            if _is_internal_error(import_result):
+                msg = 'Import failed {0} times, due to internal server error' \
+                      '; {1}'.format(number_of_attempts, import_result.text)
+                raise exceptions.DSLParsingLogicException(13, msg)
+            return import_result
+        # If any of the previously caught exceptions (and the URLRequired)
```

Comment:
a bit confusing here... I think it's clearer to explicitly state "If a ConnectionError, Timeout or URLRequired exceptions should arise, will raise a custom exception"